### ISSUE TITLE
[Bug?] Restricting input space for acqf optimization leads to visibly better acqf optima

### ISSUE THREAD

**Issue by K-Kielak** [pytorch/botorch#954]
# 🐛 Bug

Restricting available input space for acquisition function optimization, e.g., via `fixed_features` kwarg, can lead to visibly better acquisition function values. This is rather unexpected since, even in the case when `fixed_features` were chosen perfectly, the unrestricted optimization should yield at least the same results (by implicitly optimizing input to be inline with these 'perfect' `fixed_features` configuration).

## To reproduce 
Below I produced an example using `fixed_features`, but the same can be achieved with `inequality_constraints`

** Code snippet to reproduce **
```python
target = 2.1578

gp = SingleTaskGP(inp, out)
mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_model(mll)

poi_fn = qProbabilityOfImprovement(model=gp, best_f=target)
opt_acqf_kwargs = {
    "bounds": torch.stack([torch.zeros(7), torch.ones(7)]), 
    "q": 10, 
    "num_restarts": 32, 
    "raw_samples": 512, 
    "sequential": False
}

poi_no_fixed = optimize_acqf(poi_fn, **opt_acqf_kwargs)
poi_with_fixed = optimize_acqf(poi_fn, fixed_features={5: 0.0, 6: 0.0}, **opt_acqf_kwargs)
print(poi_no_fixed[1])
print(poi_with_fixed[1])
```

** Data **
`inp` operates in a 7D space normalized to a unit hypercube. The small caveat here may be that all this datapoints come with feature 5 and 6 equal to 0 (i.e., everything with greater values for feature 5 and 6 is completely unexplored).
`out` is standardized to mean 0 and standard deviation 1. Target of 2.1578 as set in the code has not been achieved yet.

## Output
`poi_no_fixed[1]` is consistently less than `poi_with_fixed[1]`. E.g:.
```
tensor(0.3959, dtype=torch.float64)
tensor(0.4410, dtype=torch.float64
```

## Expected Behavior
In most cases I'd assume POI without any restrictions on input space (i.e., without fixed features) should be greater than POI with fixed features, simply due to a larger possible exploration space. Especially that playing with features 5 and 6 introduces a lot of uncertainty (potential for improvement).

However, even in the case where indeed all the 10 points that together are best at maximizing POI had feat5 = 0 and feat6 = 0, I'd expect POI without restrictions to be at least as good as POI with restrictions. It should optimize the acquisition fn to suggest all the points with feat5 and feat6 = 0 (and hence equal acqf value).

## System information
BoTorch Version: 0.5.1
GPyTorch Version: 1.5.1
PyTorch Version: 1.9.1+cu102
Computer OS: Ubuntu 20.04.2

## Other comments
The only possible explanation I came up with is that it may be caused by some limits of MC optimization. Especially since POI is non-convex? However, if that is the case, I'd assume that playing with things such as `num_restarts` should somewhat mitigate this issue, but based on my tests it is not the case.

**Comment by Balandat** [comment]
Interesting. Your logic makes sense, mathematically the optimum over the unrestricted space should by definition be at least as high as the optimum over the restricted space. My suspicion here is that this function is particularly nasty and hard to optimize, especially since you're using `q=10` and are setting `sequential=False`, which means that you're optimizing a highly nonconvex function over a 70-dimensional space (`q*d=10*7`). 

The first thing I would suggest for such a large `q` is to use `sequential=True`, which will instead make this solve 10 7-dim problems. You can also crank up `raw_samples` to 1024 or even more to give the initialization strategy some more points to work with in the hope of not missing any local optima.

**Comment by Balandat** [comment]
So since there is no data whatsover outside the subspace `feat5=feat6=0` I am wondering if the model fits reasonably well. Can you check what the lengthscales of the resulting model are? They should be essentially at the mode of the prior since there is zero information from the data in that direction. You can just paste the `state_dict` of the model here.

**Comment by K-Kielak** [comment]
Output of `gp.state_dict()`:

```
OrderedDict([('likelihood.noise_covar.raw_noise',
              tensor([0.0948], dtype=torch.float64)),
             ('likelihood.noise_covar.noise_prior.concentration',
              tensor(1.1000, dtype=torch.float64)),
             ('likelihood.noise_covar.noise_prior.rate',
              tensor(0.0500, dtype=torch.float64)),
             ('likelihood.noise_covar.raw_noise_constraint.lower_bound',
              tensor(1.0000e-04, dtype=torch.float64)),
             ('likelihood.noise_covar.raw_noise_constraint.upper_bound',
              tensor(inf, dtype=torch.float64)),
             ('mean_module.constant', tensor([-0.1509], dtype=torch.float64)),
             ('covar_module.raw_outputscale',
              tensor(0.7460, dtype=torch.float64)),
             ('covar_module.base_kernel.raw_lengthscale',
              tensor([[-0.9082, -0.9171, -0.8953, -0.8993, -0.8992, -0.9273, -0.9273]],
       dtype=torch.float64)),
             ('covar_module.base_kernel.lengthscale_prior.concentration',
              tensor(3., dtype=torch.float64)),
             ('covar_module.base_kernel.lengthscale_prior.rate',
              tensor(6., dtype=torch.float64)),
             ('covar_module.base_kernel.raw_lengthscale_constraint.lower_bound',
              tensor(0., dtype=torch.float64)),
             ('covar_module.base_kernel.raw_lengthscale_constraint.upper_bound',
              tensor(inf, dtype=torch.float64)),
             ('covar_module.outputscale_prior.concentration',
              tensor(2., dtype=torch.float64)),
             ('covar_module.outputscale_prior.rate',
              tensor(0.1500, dtype=torch.float64)),
             ('covar_module.raw_outputscale_constraint.lower_bound',
              tensor(0., dtype=torch.float64)),
             ('covar_module.raw_outputscale_constraint.upper_bound',
              tensor(inf, dtype=torch.float64))])
```

**Comment by Balandat** [comment]
These look reasonable. What I could see happening here is that as you move off the subspace (with the other dimensions fixed) the posterior mean drops faster that the uncertainty increases and thus PoI overall drops. Might be good to see if that's what's happening by computing means/vars on some small norm ball around a point on the subspace.
